### PR TITLE
REDUCE: Fix overflow in short_sum tests

### PR DESCRIPTION
**Pull request by brminich**
Fix type overflow in `shmem_short_sum_to_all` tests.

Fixes openucx/ucx#6547

### CODE REVIEW

**Comment by brminich** [comment]
@hoopoepg, pls review

**Review by hoopoepg — APPROVED** [review]
it seems coding guidelines are not applicable here... then ok :)

File: verifier/reduce/osh_reduce_tc27.c [hunk 1]
```diff
@@ -925,8 +966,12 @@ static int test_item8(void)
...
                 else
                 {
-                    int k = num_proc;
-                    while (k--) expect_value += (k % 2 ? 0 : k);
+                    int k = sys_min(max_proc_num, num_proc - 1);
+                    while (k)
+                    {
+                        expect_value += (k % 2 ? 0 : k);
+                        --k;
+                    }
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:974]: it seems pretty complicated, either need a good comment to explain or just don't check data if an overlow is possible (e.g num_proc*num_proc > SHRT_MAX)

File: verifier/reduce/osh_reduce_tc27.c [hunk 2]
```diff
@@ -77,6 +79,23 @@ static const AOPT_DESC  self_opt_desc[] =
 };
 
 
+int osh_reduce_max_n(TYPE_VALUE max_value, int count_even)
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:82]: what is this function doing? if we keep it need a comment or better name..
also, can't we calculate this value without a loop? AFAIR there is a formula for sum of series

File: verifier/reduce/osh_reduce_tc27.c [hunk 3]
```diff
@@ -77,6 +79,41 @@ static const AOPT_DESC  self_opt_desc[] =
 };
 
 
+
+ /*
+  * Solve quadratic equation to find the closest value of n for the given sum
+ * of first n numbers, according to the formulas:
+ * sum = n * (n + 1) / 2 - for the sum of all numbers
+ * sum = n * (n + 2) / 4 - for the sum of even numbers
+ */
```
Comment (hoopoepg) [inline verifier/reduce/osh_reduce_tc27.c:88]: unaligned comments

**Review by hoopoepg — APPROVED** [review]
just minor unaligned comments :)

**Review by hoopoepg — APPROVED** [review]

**Comment by brminich** [comment]
@yosefe, can you please review?

File: verifier/reduce/osh_reduce_tc27.c [hunk 4]
```diff
@@ -77,6 +79,41 @@ static const AOPT_DESC  self_opt_desc[] =
...
+        b = 1;
+        c = -2 * sum;
+    }
+
+    discriminant = b*b - 4*a*c;
+    if (discriminant < 0) {
+        return 0;
+    }
+
+    x = -b  + sqrt(discriminant);
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:111]: - extra space
- can remove "x"

File: verifier/reduce/osh_reduce_tc27.c [hunk 5]
```diff
@@ -77,6 +79,41 @@ static const AOPT_DESC  self_opt_desc[] =
...
+        b = 2;
+        c = -4 * sum;
+    }
+    else
+    {
+        b = 1;
+        c = -2 * sum;
+    }
+
+    discriminant = b*b - 4*a*c;
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:106]: discriminant = (b * b) - (4 * a * c);

File: verifier/reduce/osh_reduce_tc27.c [hunk 6]
```diff
@@ -77,6 +79,41 @@ static const AOPT_DESC  self_opt_desc[] =
...
+    if (discriminant < 0) {
+        return 0;
+    }
+
+    x = -b  + sqrt(discriminant);
+
+    return x / (2*a);
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:113]: `return (-b + srqt(d)) / (2 * a)`

File: verifier/reduce/osh_reduce_tc27.c [hunk 7]
```diff
@@ -77,6 +79,41 @@ static const AOPT_DESC  self_opt_desc[] =
...
 
 
+
+/*
+ * Solve quadratic equation to find the closest value of n for the given sum
+ * of first n numbers, according to the formulas:
+ * sum = n * (n + 1) / 2 - for the sum of all numbers
+ * sum = n * (n + 2) / 4 - for the sum of even numbers
+ */
+int osh_reduce_max_n(TYPE_VALUE sum, int count_even)
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:89]: static

File: verifier/reduce/osh_reduce_tc27.c [hunk 8]
```diff
@@ -301,8 +346,8 @@ static int test_item2(void)
         /* Define expected value */
         expect_value = 0;
         {
-            int k = num_proc;
-            while (k--) expect_value += k;
+            int k = sys_min(max_proc_num, num_proc - 1);
+            while (k) expect_value += k--;
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:350]: maybe rewrite as a for loop?
```
for (k = sys_min(max_proc_num, num_proc - 1), expect_value = 0; k > 0; --k) {
   expect_value += k;
}
```

File: verifier/reduce/osh_reduce_tc27.c [hunk 9]
```diff
@@ -473,8 +526,8 @@ static int test_item4(void)
...
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:530]: unite code with test_item2() ?

File: verifier/reduce/osh_reduce_tc27.c [hunk 10]
```diff
@@ -677,8 +739,8 @@ static int test_item6(void)
                 /* Define expected value */
                 expect_value = 0;
                 {
-                    int k = num_proc;
-                    while (k--) expect_value += k;
+                    int k = sys_min(max_proc_num, num_proc - 1);
+                    while (k) expect_value += k--;
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:743]: unite code with test_item2() ?

File: verifier/reduce/osh_reduce_tc27.c (hunk 1 above)
Comment (yosefe) [reply verifier/reduce/osh_reduce_tc27.c:1017]: unite code with test_item7?

File: verifier/reduce/osh_reduce_tc27.c [hunk 11]
```diff
@@ -925,8 +1009,12 @@ static int test_item8(void)
                 if (my_proc % 2)    expect_value = DEFAULT_VALUE;
                 else
                 {
-                    int k = num_proc;
-                    while (k--) expect_value += (k % 2 ? 0 : k);
+                    int k = sys_min(max_proc_num, num_proc - 1);
+                    while (k)
+                    {
+                        expect_value += (k % 2 ? 0 : k);
```
Comment (yosefe) [inline verifier/reduce/osh_reduce_tc27.c:1015]: `expect_value += (k % 2) ? 0 : k;`

File: verifier/reduce/osh_reduce_tc27.c (hunk 9 above)
Comment (brminich) [reply verifier/reduce/osh_reduce_tc27.c:530]: that is not common for verifier, all tests have maany common code, but are written as kinda stand-alone tests

**Comment by brminich** [comment]
@yosefe fixed

**Review by yosefe — APPROVED** [review]

**Comment by brminich** [comment]
@yosefe, squashed

**Comment by brminich** [comment]
@yosefe, ok to merge?

**Comment by brminich** [comment]
@yosefe, ok to merge?

**Pull request merged**